# Семинар 2. Совсем краткое введение в Питон и задачка про ёжиков:)



Дата: 26.09.2016
Конспект: Валерия Жувагина

## Печатаем текст в Markdown

Немного о том, как писать здесь формулы и текст.Например, мы пишем: 
Решаем уравнение $x+7=8$. Или же дробь: 

$$
\frac{45}{56}=7
$$

Формула стала красивой. Пишем формулы в уже знакомом многим TeXe. Для того, чтобы ячейку перевести из режима редактирования в готовый текст, нужно нажать shift+ENTER а чтобы её снова можно было редактировать, нужно покликать на текст :) Каждая ячейка может быть либо типа code (чтобы написать код), либо типа markdown (чтобы напечатать текст). Тип ячейки можно выбрать в toolbar наверху. Заголовки пишутся так: # Заголовок ## Подзаголовок

Мы обсудили, как писать текст, теперь  о том, как писать команды для компьютера, которые он должен выполнять.

## Code и пакеты

Выбираем в toolbar вместо "Markdown" "Code" и решаем, например, вот такую, достаточно примитивную, задачу

In [3]:
5 + 9

14

То есть, если мы считаем вероятность, то мы её можем считать прямо так в тексте. Тут и графики строить можно и на другой язык программирования переходить!

Какие могут потребоваться команды, с чем мы работаем часто при расчёте вероятностей:
Нужно понимать, что Питон сам по себе мало что умеет, но на Питоне написано много библиотек, дополнительных пакетов. Наша задача, найти пакет, в котором находится необходимая нам функция и ей воспользоваться. Самому писать цикл не надо, хоть уметь это делать хорошо. 
Пакеты с которыми мы будем работать: numpy, scipy (для научных вычислений); seabourn и ggplot (для графиков); matplotlib; для работы с табличками данных pandas; scikit.learn для оценки моделей.
Нужно вам, например, посчитать факториал. Гуглите, где живёт эта функция и находите пакет.
Если мы не знаем, где она живет то гуглим: how to calculate factorial in python, а если знаем, то гуглим factorial scipy :)
Находим необходимый модуль и пишем в ячейке import и название модуля. Например:

In [4]:
import scipy.misc

In [5]:
scipy.misc.factorial(7)

array(5040.0)

Это мы попросили Питон посчитать 7!

Можно делать по другому:

In [6]:
from scipy.misc import *

In [8]:
factorial(7)

array(5040.0)

Функция choose k from n есть и в numpy, и в scipy. Можно найти либо под названием choose(), либо comb().

Теперь решим задачку про ежей:

# Ежи (Листок 1, задача 12)

В каждой вершине треугольника по ёжику. Каждую минуту с вероятностью 0,7 каждый ёжик независимо от других двигается по часовой стрелке и с вероятностью 0,3 против часовой. Обозначим T - время до встречи всех ежей в одной вершине. Найдите P(T=1), P(T=2), P(T=3), E(T).

Итак, наше изначальное состояние-все ежи в разных углах. Это конфигурация 1-1-1

![](http://s018.radikal.ru/i523/1610/a0/a9d6688688e3.png)

Конфигурации будем считать по часовой от большего количества ежей в углу к меньшему.

Поскольку вероятность того, что ёж пойдет по часовой не равна вероятности того, что ёж пойдет против часовой, конфигурация 2-1-0 не равна конфигурации 2-0-1.

Возможные конфигурации: 1-1-1, 2-1-0, 2-0-1, 3-0-0

![](http://s014.radikal.ru/i328/1610/d4/bfcbb5a62a28.png)

Вот так выглядят наши возможные переходы из одной конфигурации в другую или саму в себя:

![Вот так выглядят наши возможные переходы из одной конфигурации в другую или саму в себя:](http://i053.radikal.ru/1610/ab/1f380b13393a.png)



## Пункт 1. Найти P(T=1)


Как мы видим, сразу перейти из конфигурации 1-1-1 в конфигурацию 3-0-0 невозможно, нужно сделать как минимум два шага. Поэтому P(T=1)=0



## Пункт 2. Найти P(T=2)

Посчитаем вероятность перехода из 1-1-1 в 2-0-1. Тут всегда два ежа должны пойти против часовой, а один по часовой и возможных комбинаций 3, поэтому:

$$P(\text{из 1-1-1 в 2-0-1})=3\times0.3\times0.3\times0.7$$

In [3]:
0.3*0.3*0.7

0.063

In [16]:
3*0.063

0.189

Для того, чтобы из 2-0-1 ёжики перешли в 3-0-0, нужно, чтобы два из них пошли по часовой (те, которые уже встретились), а один пошёл против часовой. Соответственно, вероятность перехода из 2-0-1 в 3-0-0:

$$P(\text{из 2-0-1 в 3-0-0}) =0.7\times0.7\times0.3$$



In [6]:
0.3*0.7*0.7

0.147

Значит, перемножение вероятностей даст нам все исходы, при которых мы из 1-1-1 через 2-0-1 попадаем в 3-0-0


In [17]:
0.147*0.189

0.027783

В случае с 2-1-0 всё наоборот: чтобы попасть в него нам нужно, чтобы два ежа пошли по часовой и один против. А чтобы из 2-1-0 попасть в 3-0-0 нам нужно, чтобы 2 ежа пошли против часовой, а один по. 

In [9]:
0.7*0.7*0.3*3


0.4409999999999999

In [13]:
0.3*0.3*0.7

0.063

In [18]:
0.063*0.441

0.027783

Вероятность того, что ёжики встретятся за два хода=сумма вероятностей того, что они перейдут из 1-1-1 в 2-0-1, а потом в 3-0-0
и того, что они перейдут из 1-1-1 в 2-1-0, а потом в 3-0-0

In [10]:
0.027783+0.027783

0.055566

$$Ответ: P(T=2)=0.055566$$

## Пункт 3. Найти P(T=3)

Есть 6 вариантов, при которых мы делаем 3 шага до конца игры. 

![](http://i013.radikal.ru/1610/33/6c12dbe667b4.png)

1 и 2 вариант: 1-1-1 переходит сам в себя, потом как в T=2 (либо в 2-0-1, либо в 2-1-0)

In [23]:
0.37*0.441*0.063

0.010279710000000001

In [24]:
0.37*0.189*0.147

0.010279710000000001

In [25]:
0.01027971*2

0.02055942

3 и 4 вариант. Переходим либо в 2-0-1, либо в 2-1-0, эта конфигурация переходит саму в себя, а потом приходим в 3-0-0

In [12]:
0.189*0.37*0.147

0.010279710000000001

In [11]:
0.441*0.37*0.063

0.010279710000000001

In [29]:
0.01027971*2

0.02055942

5 и 6 вариант: переходим либо в 2-0-1, либо в 2-1-0, затем, если мы попали в 2-0-1, она переходит в 2-1-0, а если мы попали в 2-1-0, она переходит в 2-0-1. Потом, как обычно, переходим в 3-0-0.

In [32]:
0.441*0.441*0.147

0.012252303

In [31]:
0.189*0.189*0.063

0.002250423

In [33]:
0.012252303+0.002250423

0.014502726

Складываем все возможные варианты:

In [35]:
0.02055942+0.02055942+0.014502726

0.055621566

$$Ответ: P(T=3)=0.055621566$$

## Пункт 4. Найти E(T)

Предположим, что из точки 1-1-1 в среднем надо ждать $a$ ходов до встречи, из 2-0-1 в среднем $b$ ходов до встречи, а из 2-1-0 в среднем $c$ ходов до встречи.

Рассмотрим ситуацию 1-1-1. Мы можем потратить один ход и перейти в неё же. Запишем это как $0.37 (1+a)$. 
Мы также можем потратить один ход и перейти туда, где нам в среднем надо ждать $b$ ходов, то есть в ситуацию 2-0-1. Это будет выглядеть как: $0.189(b+1)$, а если же мы перейдем в 2-1-0, то: $0.441(1+с)$.

По такому же принципу распишем ситуации, где нам осталось ждать $b$ и $c$ ходов (то есть 2-0-1 и 2-1-0). У нас получится следующая система уравнений:

\begin{equation*}
 \begin{cases}
  a=0.37(1+a)+0.189(1+b)+0.441(1+c)
   \\
   b=0.294(1+a)+0.37(1+b)+0.189(1+c)+0.147
   \\
   c=0.126(1+a)+0.441(1+b)+0.37(1+c)+0.063
 \end{cases}
\end{equation*}

Ожидаемое количество ходов-это и есть $a$, то есть ожидаемое количество ходов в точке начала игры. Решение данной системы уравнений даёт результат: $a=E(T)=14\frac{2}{7}$

$$Ответ: E(T)=14\frac{2}{7}$$

![](http://s018.radikal.ru/i512/1610/8f/5c310d7af7c2.jpg)